### PR TITLE
Skyline/work/20220624 redundant spectra dropdown view spectral libraries

### PR DESCRIPTION
**Pull request by aaronigra**
Support for a dropdown in the view spectral libraries menu to view redundant spectra matches in the given library.

### CODE REVIEW

File: pwiz_tools/Skyline/SettingsUI/ViewLibraryDlg.cs [hunk 1]
```diff
@@ -633,6 +641,35 @@ private void UpdateListPeptide(int selectPeptideIndex)
             }
         }
 
+
+        public class comboOption
+        {
+            public SpectrumInfoLibrary library;
```
Comment (nickshulman) [inline pwiz_tools/Skyline/SettingsUI/ViewLibraryDlg.cs:647]: This field should be called "spectrumInfo", or maybe "spectrumInfoLibrary". 

A "SpectrumInfoLibrary" is a type of "SpectrumInfo", not a type of "Library".

File: pwiz_tools/Skyline/SettingsUI/ViewLibraryDlg.cs [hunk 2]
```diff
@@ -695,8 +742,8 @@ public void UpdateUI(bool selectionChanged = true)
 
                 if (-1 != index)
                 {
-                    SpectrumPeaksInfo spectrum;
-                    if (_selectedLibrary.TryLoadSpectrum(_peptides[index].Key, out spectrum))
+                    SpectrumPeaksInfo isLoadableSpectrum;
```
Comment (brendanx67) [inline pwiz_tools/Skyline/SettingsUI/ViewLibraryDlg.cs:745]: Usually, a name like "is..." would be reserved for a boolean value. Maybe something like loadedSpectrum?

File: pwiz_tools/Skyline/SettingsUI/ViewLibraryDlg.cs [hunk 3]
```diff
@@ -774,22 +835,23 @@ public void UpdateUI(bool selectionChanged = true)
                         graphControl.IsEnableVPan = graphControl.IsEnableVZoom =
                                                     !Settings.Default.LockYAxis;
                         // Update file and retention time indicators
-                        var bestSpectrum = _selectedLibrary.GetSpectra(_peptides[index].Key,
-                            IsotopeLabelType.light, LibraryRedundancy.best).FirstOrDefault();
-                        if (bestSpectrum != null)
+                        // var bestSpectrum = _selectedLibrary.GetSpectra(_peptides[index].Key,
```
Comment (brendanx67) [inline pwiz_tools/Skyline/SettingsUI/ViewLibraryDlg.cs:838]: Not sure we need the old code in a comment here.

File: pwiz_tools/Skyline/SettingsUI/ViewLibraryDlg.cs [hunk 4]
```diff
@@ -875,6 +937,35 @@ public void UpdateUI(bool selectionChanged = true)
...
+            else
+            {
+                comboRedundantSpectra.Items.Clear();
+                options.Sort();
+                foreach (var opt in options)
+                {
+                    comboRedundantSpectra.Items.Add(opt);
+                    if (opt.spectrumInfoLibrary.IsBest)
+                    {
+                        comboRedundantSpectra.SelectedIndexChanged -= redundantSpectrum_changed;
```
Comment (brendanx67) [inline pwiz_tools/Skyline/SettingsUI/ViewLibraryDlg.cs:956]: Worth adding a comment here about what you are doing, i.e. avoid updating the UI.

File: pwiz_tools/Skyline/TestTutorial/TargetedMSMSTutorialTest.cs [hunk 5]
```diff
@@ -145,7 +146,6 @@ private void LowResTestPartOne(RefinementSettings.ConvertToSmallMoleculesMode as
             AssertEx.IsDocumentState(document, null, 3, expectedMoleculeCount, expectedTransitionGroupCount,
                 expectedTransitionCount);
             CheckConsistentLibraryInfo();
-
```
Comment (brendanx67) [inline pwiz_tools/Skyline/TestTutorial/TargetedMSMSTutorialTest.cs:148]: Doesn't seem like a relevant change.

File: pwiz_tools/Skyline/TestTutorial/TargetedMSMSTutorialTest.cs [hunk 6]
```diff
@@ -458,6 +458,7 @@ private void LowResTest()
             var allChromGraph = WaitForOpenForm<AllChromatogramsGraph>();
             RunUI(() => allChromGraph.Left = SkylineWindow.Right + 20);
             PauseForScreenShot<AllChromatogramsGraph>("Loading chromatograms window", 19);
+
```
Comment (brendanx67) [inline pwiz_tools/Skyline/TestTutorial/TargetedMSMSTutorialTest.cs:461]: Doesn't seem like a relevant change.

File: pwiz_tools/Skyline/TestTutorial/TargetedMSMSTutorialTest.cs [hunk 7]
```diff
@@ -1009,5 +1011,41 @@ private static double[] AreaGraphDotProducts
                 return dotpActuals;
             }
         }
+
+        //Tests the redundant spectrum dropdown menu in the ViewSpectralLibraries dialogue
```
Comment (brendanx67) [inline pwiz_tools/Skyline/TestTutorial/TargetedMSMSTutorialTest.cs:1015]: Use a C# header comment more like this: 

/// &lt;summary>
/// Tests the redundant spectrum dropdown menu in the &lt;see ref="ViewLibraryDlg" /> form
/// &lt;/summary>

File: pwiz_tools/Skyline/TestTutorial/TargetedMSMSTutorialTest.cs [hunk 8]
```diff
@@ -1009,5 +1011,41 @@ private static double[] AreaGraphDotProducts
                 return dotpActuals;
             }
         }
+
+        //Tests the redundant spectrum dropdown menu in the ViewSpectralLibraries dialogue
+        private void TestRedundantComboBox()
+        {
+            var dlg = ShowDialog<ViewLibraryDlg>(SkylineWindow.ViewMenu.ViewSpectralLibraries);
```
Comment (brendanx67) [inline pwiz_tools/Skyline/TestTutorial/TargetedMSMSTutorialTest.cs:1018]: Add some comments below about what you are testing.

File: .gitignore [hunk 9]
```diff
@@ -448,3 +448,11 @@ SkylineTester Results
 /BrowseOnly
 /x64/BrowseOnly
 /qb64.bat
+/pwiz_aux/msrc/utility/vendor_api/Bruker/BAF2SQL-README.txt
```
Comment (brendanx67) [inline .gitignore:451]: You just mentioned this, but you should not need to alter .gitignore for the project.

**Review by brendanx67 — CHANGES_REQUESTED** [review]
Looking good. My requested changes should be pretty quick.

File: pwiz_tools/Skyline/SettingsUI/ViewLibraryDlg.cs [hunk 10]
```diff
@@ -738,8 +785,22 @@ public void UpdateUI(bool selectionChanged = true)
                         }
                         adducts.AddRange(showAdducts.Where(a => charges.Contains(Math.Abs(a.AdductCharge)) && !adducts.Contains(a))); // And the unranked charges as well
 
-                        var spectrumInfo = _selectedLibrary.GetSpectra(_peptides[index].Key, null, LibraryRedundancy.best).FirstOrDefault();
-                        var spectrumInfoR = LibraryRankedSpectrumInfo.NewLibraryRankedSpectrumInfo(spectrum,
+                        //Get all redundant spectrum for the selected peptide, add them to comboRedundantSpectra
```
Comment (bspratt) [inline pwiz_tools/Skyline/SettingsUI/ViewLibraryDlg.cs:788]: Totally nitpicking here, but our commenting style has a leading space (and, good, capital letter)

Comment (bspratt) [reply pwiz_tools/Skyline/SettingsUI/ViewLibraryDlg.cs:788]: (But, a wise coder once told me that "The only proper code formatting style is the one that's already in use." The idea is not to trip up the eye of whoever is digging through your code looking for who-knows-what.)

**Review by bspratt — DISMISSED** [review]
This looks good though I would encourage you to go through and make your comments look like all the others already there. It's not busy work,. I promise you.

**Comment by bspratt** [comment]
Also make sure you deal with ReSharper code inspection warnings, if you haven't already. You can run those locally, and they also happen on TeamCity https://teamcity.labkey.org/viewLog.html?buildId=1881735&tab=Inspection&buildTypeId=ProteoWizard_WindowsX8664msvcProfessionalSkylineResharperChecks

**Comment by brendanx67** [comment]
One last thought: You should be using ComboHelper.AutoSizeDropDown() when you change the contents of the dropdown list. That way a user doesn't have to resize the entire form just to see the full contents of the dropdown list when the file names are very long.

**Comment by brendanx67** [comment]
Okay. Looks very close to me now after a bit of clean-up. Let's go over the core function to update the combo box, and then we should be able to merge it tomorrow.

**Review by brendanx67 — APPROVED** [review]

**Pull request merged**